src/hotspot/share/utilities/lockFreeQueue.inline.hpp

```diff
@@ -0,0 +1,177 @@
+
+template<typename T, T* volatile* (*next_ptr)(T&)>
+void LockFreeQueue<T, next_ptr>::set_next(T& node, T* new_next) {
+    Atomic::store(next_ptr(node), new_next);
+}
+
+template<typename T, T* volatile* (*next_ptr)(T&)>
+LockFreeQueue<T, next_ptr>::LockFreeQueue() : _head(NULL), _tail(NULL) {}
+
+#ifdef ASSERT
+template<typename T, T* volatile* (*next_ptr)(T&)>
+LockFreeQueue<T, next_ptr>::~LockFreeQueue() {
+  assert(_head == NULL, "precondition");
+  assert(_tail == NULL, "precondition");
+}
+#endif
+
+template<typename T, T* volatile* (*next_ptr)(T&)>
+T* LockFreeQueue<T, next_ptr>::top() const {
+  return Atomic::load(&_head);
+}
+
+template<typename T, T* volatile* (*next_ptr)(T&)>
+size_t LockFreeQueue<T, next_ptr>::length() const {
+  size_t result = 0;
+  for (const T* current = top(); current != NULL; current = next(*current)) {
+    ++result;
+  }
+  return result;
+}
+
+// An append operation atomically exchanges the new tail with the queue tail.
+// It then sets the "next" value of the old tail to the head of the list being
+// appended; it is an invariant that the old tail's "next" value is NULL.
+// But if the old tail is NULL then the queue was empty.  In this case the
+// head of the list being appended is instead stored in the queue head; it is
+// an invariant that the queue head is NULL in this case.
+//
+// This means there is a period between the exchange and the old tail update
+// where the queue sequence is split into two parts, the list from the queue
+// head to the old tail, and the list being appended.  If there are concurrent
+// push/append operations, each may introduce another such segment.  But they
+// all eventually get resolved by their respective updates of their old tail's
+// "next" value.  This also means that pop operations must handle an object
+// with a NULL "next" value specially.
+//
+// A push operation is just a degenerate append, where the object being pushed
+// is both the head and the tail of the list being appended.
+template<typename T, T* volatile* (*next_ptr)(T&)>
+void LockFreeQueue<T, next_ptr>::append(T& first, T& last) {
+  assert(next(last) == NULL, "precondition");
+  T* old_tail = Atomic::xchg(&_tail, &last);
+  if (old_tail == NULL) {       // Was empty.
+    Atomic::store(&_head, &first);
+  } else {
+    assert(next(*old_tail) == NULL, "invariant");
+    set_next(*old_tail, &first);
+  }
+}
+
+template<typename T, T* volatile* (*next_ptr)(T&)>
+template<bool use_rcu>
+T* LockFreeQueue<T, next_ptr>::pop() {
+  while (true) {
+    // Use a critical section per iteration, rather than over the whole
+    // operation. We're not guaranteed to make progress. Lingering in one
+    // CS could defer the write-side operation of RCU synchronization
+    // too long, leading to unwanted effect. E.g., if the write-side
+    // returns released objects to a free list for reuse, it could cause
+    // excessive allocations.
+    GlobalCounter::ConditionalCriticalSection<use_rcu> cs(use_rcu ?
```

Comment:
` GlobalCounter::ConditionalCriticalSection<use_rcu> cs(Thread::current());` should be fine, not sure how much is gained by skipping the `Thread::current()` call.

Reply:
Superceded by not adding the ConditionalCriticalSection.